### PR TITLE
Implement Sniffer API

### PR DESCRIPTION
**Pull request by Frostie314159**
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [x] My changes are in accordance to the [esp-rs API guidelines](https://github.com/esp-rs/esp-hal/blob/main/API-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/CONTRIBUTING.md) and followed its instructions.

#### Description
This PR adds support for a sniffer API, which allows the reception and transmission of raw 802.11 frames.

#### Testing
I've created two examples for this. One which transmits a beacon every 100ms and another, which listens for beacon frames.

### CODE REVIEW

**Comment by Frostie314159** [comment]
I added an example for transmission, which just emits a beacon frame every 100ms. @bjoernQ I used my library for serialization, I hope that's ok.
I'm not sure that there exists any sensible way, to make an async API for this.

**Comment by leighleighleigh** [comment]
re: async, I assumed I could just store the received frames in a queue - are you thinking the large volume of frames would make that approach infeasible? Or is there something else which makes async difficult? :) thanks

**Comment by Frostie314159** [comment]
I thought about storing them in a queue, but that'd require massive amounts of stack space.

File: esp-wifi/src/wifi/mod.rs [hunk 1]
```diff
@@ -1770,10 +1780,208 @@ fn convert_ap_info(record: &include::wifi_ap_record_t) -> AccessPointInfo {
     }
 }
 
+#[cfg(not(any(esp32c6)))]
```
Comment (MabezDev) [inline esp-wifi/src/wifi/mod.rs:1783]: I know it's not related to this PR, but @bjoernQ do you know why this is unavailable on the C6?

**Review by MabezDev — APPROVED** [review]
This is really neat, thanks for the PR!

Regarding async, I think we can figure this out in another PR if we decide its something important.

LGTM

**Comment by Frostie314159** [comment]
I actually had an implementation for async drafted up, but this would be incompatible with the sync implementation and would require copying the packet at least once.

**Comment by MabezDev** [comment]
> I actually had an implementation for async drafted up, but this would be incompatible with the sync implementation and would require copying the packet at least once.

We can explore this in a follow up PR, perhaps we can find a way around the copy.

**Comment by Frostie314159** [comment]
We should definitely explore that. 
My approach was, to have a static `AtomicWaker` and a `Mutex<Option<*mut [u8]>>`. If the pointer is `Some`, we copy the packet over and wake the waker. 
The current promiscuous API gives you a pointer, to a buffer allocated by the Wi-Fi driver, which is freed after the callback is called. See `ppProcessRxPktHdr` in `libpp.a`.

**Review by bjoernQ — APPROVED** [review]
LGTM - Thanks - really nice addition

**Pull request merged**